delete() calls to a report file will move it into sdcard storage

crankycoder opened the pull request:
This is a patch to get #1255 in so that delete() calls to a report file will actually move it into sdcard storage.

Most of the code here is actually robolectric related test code.  Important bits to note are:
- There is a test that exercises the checkbox in the DeveloperActivityFragment and setting the pref
- There is a test that makes sure that the SDCARD is mounted
  - the sdcard may be in an unmounted state, but robolectric will let you write to it anyway.  There doesn't seem to be a good way to test or mock this out.
- There is a test to verify the toast string
- There is a test to make sure files are moved from the report directory to the sdcard archive path.

crankycoder committed:
Added a developer option to save stumble logs

crankycoder committed:
added a basic unit test to check that the developer option for saving stumble logs works.

crankycoder committed:
There's no good reason to *not* use .equals() when comparing the filename to the sentinel value of MEMORY_BUFFER_NAME

crankycoder committed:
Create a toast if the sdcard is unavailable and force the checkbox
to be unchecked

crankycoder committed:
added sdcard mounted and unmounted tests when checking save logs in dev options.

crankycoder committed:
added test to show that files get moved into the sdcard_archive_path on delete()

garvankeeley commented on android/src/main/java/org/mozilla/mozstumbler/client/subactivities/DeveloperActivity.java:
```diff
@@ -6,6 +6,7 @@
 import android.app.AlertDialog;
 import android.content.DialogInterface;
 import android.os.Bundle;
+import android.os.Environment;
```
not needed

garvankeeley commented on android/src/main/java/org/mozilla/mozstumbler/client/subactivities/DeveloperActivity.java:
```diff
@@ -13,23 +14,24 @@
 import android.widget.CompoundButton;
-import android.widget.Spinner;
 import android.widget.TextView;
+import android.widget.Toast;
```
not needed

garvankeeley commented on android/src/main/java/org/mozilla/mozstumbler/client/subactivities/DeveloperActivity.java:
```diff
@@ -13,23 +14,24 @@
+
+import java.io.File;
+
+import static org.mozilla.mozstumbler.R.string.create_log_archive_failure;
```
28-30 not needed

garvankeeley commented on android/src/main/java/org/mozilla/mozstumbler/service/stumblerthread/datahandling/DataStorageManager.java:
```diff
@@ -20,6 +25,8 @@
 import java.util.TimerTask;
 
+import static org.mozilla.mozstumbler.R.string.create_log_archive_failure;
+
```
No resources in the service per favor

garvankeeley commented on android/src/main/java/org/mozilla/mozstumbler/service/stumblerthread/datahandling/DataStorageManager.java:
```diff
@@ -227,6 +239,34 @@ private String getStorageDir(Context c) {
 
+
+    public static boolean archiveDirCreatedAndMounted(Context ctx) {
+        File saveDir = new File(sdcard_archive_path());
```
Can we pull this archiveDirCreatedAndMounted out to a client-side interface. maybe a subclass. Same with sdcard_archive_path, which should be sdcardArchivePath

crankycoder replied:
I've shuffled these bits into the DeveloperActivity now.

garvankeeley commented on android/src/main/java/org/mozilla/mozstumbler/service/stumblerthread/datahandling/DataStorageManager.java:
```diff
@@ -513,14 +538,117 @@ public void writeSyncStats(long time, long bytesSent, long totalObs, long totalC
+
+        // delete the original file
+        ok = ok & aFile.delete();
+        return ok;
```
return ok && aFile.delete()?

crankycoder replied:
![doh](http://i2.kym-cdn.com/entries/icons/original/000/011/786/homer-simpson-doh.jpg)

garvankeeley commented on android/src/main/java/org/mozilla/mozstumbler/service/stumblerthread/datahandling/DataStorageManager.java:
```diff
@@ -513,14 +538,117 @@ public void writeSyncStats(long time, long bytesSent, long totalObs, long totalC
     }
 
+    /* Pass filename returned from dataToSend() */
+    public synchronized boolean delete(String filename) {
```
Can we pull out this new logic to a client-side subclass?

garvankeeley commented:
We should go over how are these files can become useful after not deleting them.

crankycoder committed:
Optimized imports in DeveloperActivity

crankycoder committed:
* Renamed sdcard_archive_path to use camellias
* Moved generation of a toast for sdcard failure and the sdcardArchivePath methods into the DeveloperActivity

crankycoder committed:
Corrected bitwise and to logical and.  Damn you java.

crankycoder commented:
Hrm.  Subclassing DataStorageManager is a bit tricky, I'm going to need your help with this one because of the static createInstance calls.

crankycoder committed:
* split out a ClientDataStorageManager and used it in
  ClientStumberService so that we can access it from the standalone
  Mozilla Stumbler
* Moved all UI related code out of DataStorageManager and into
  ClientDSM
* Updated testcases for this adjustment
* added a toast to indicate where logs are saved to when save option
  is enabled.

crankycoder commented:
This patch is ready, but it conflicts with dev - I'm going  to hold out on doing the 3-way merge until the low battery patches land.

crankycoder committed:
Merge branch 'dev' into features/1255-move-logs

Conflicts:
	android/build.gradle
	android/src/main/java/org/mozilla/mozstumbler/client/ClientStumblerService.java
	android/src/main/java/org/mozilla/mozstumbler/client/subactivities/DeveloperActivity.java
	android/src/main/java/org/mozilla/mozstumbler/service/Prefs.java
	android/src/main/java/org/mozilla/mozstumbler/service/stumblerthread/datahandling/DataStorageManager.java
	android/src/main/res/layout/fragment_developer_options.xml
	android/src/main/res/values/strings.xml
	android/src/test/java/org/mozilla/mozstumbler/service/stumblerthread/datahandling/DataStorageManagerTest.java

crankycoder committed:
Merge branch 'dev' into features/1255-move-logs
